# Learn
In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The original dataset isn't available on Kaggle anymore, but you can find it here.

We've made a few modifications from the original dataset:

We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment.
We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with).
Data dictionary for the dataset:

- dateCrawled - When the ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing.
- price - The listed selling price of the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The type of vehicle.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The type of transmission.
- powerPS - The power of the car in PS.
- model - The car model name.
- odometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the dataset and perform some initial analysis on it. While working on this project, you'll become familiar with some of the unique benefits the JupyterLab environment has to offer for pandas.

Let's start by importing the required libraries, then load the dataset using pandas.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("raw_data/P3/autos.csv", encoding= "latin1")
autos.head()
autos.info()
autos.columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

# Observations
There are 20 columns where 5 of the columns have null values in them. These will be dealt with later.
Most of the data types are object type, some are int64.

There are 50,000 entries. 

In [3]:
renames_dictionary = {
    'dateCrawled' : 'date_crawled', 
    'name' : 'name', 
    'seller' : 'seller', 
    'offerType' : 'offer_type', 
    'price' : 'price', 
    'abtest' : 'abtest',
    'vehicleType' : 'vehicle_type', 
    'yearOfRegistration' : 'registration_year', 
    'gearbox' : 'gearbox', 
    'powerPS' : 'power_ps', 
    'model' : 'model',
    'odometer' : 'odometer', 
    'monthOfRegistration' : 'registratioon_month', 
    'fuelType' : 'fuel_type', 
    'brand' : 'brand',
    'notRepairedDamage' : 'unrepaired_damange', 
    'dateCreated' : 'ad_created', 
    'nrOfPictures' : 'nr_of_pictures', 
    'postalCode' : 'postal_code',
    'lastSeen' : 'last_seen'
}

autos = autos.rename(columns = renames_dictionary)
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   date_crawled         50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offer_type           50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicle_type         44905 non-null  object
 7   registration_year    50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   power_ps             50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  registratioon_month  50000 non-null  int64 
 13  fuel_type            45518 non-null  object
 14  brand                50000 non-null  object
 15  unrepaired_damange   40171 non-null  object
 16  ad_c

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registratioon_month fuel_type  \
0    manuell       158  andere  150,000km                    3       lpg   
1  automatik       286     7er  150,000km                    6    benzin   
2    manuell       102    golf   70,000km                    7    benzin   
3  automatik        71  fortwo   70,000km                    6    benzin   
4    manuell         0   focus  150,000km                    7    benzin   

        brand unrepaired_damange           ad_created  nr_of_pictures  \
0     peugeot               nein  2016-03-26 00:00:00               0   
1         bmw               nein  2016-04-04 00:00:00               0   
2  volkswagen               nein  2016-03-26 00:00:00               0   
3       smart               nein  2016-03-12 00:00:00               0   
4        ford               nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Changes made
We made a few changes to the column names, I believe the best way to name the columns is to use underscores instead of spaces, and snake case. Since making it all lower case would convert from the camel case to the sname case but readibility would suffer due to no spaces between the words. 

In [4]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registratioon_month fuel_type       brand  \
count       50000         50000.000000     45518       50000   
unique         13                  NaN         7          40   
top     150,000km                  NaN    benzin  volkswagen   
freq        32424                  NaN     30107       10687   
mean          NaN             5.723360       NaN         NaN   
std           NaN             3.711984       NaN         NaN   
min           NaN             0.000000       NaN         NaN   
25%           NaN             3.000000       NaN         NaN   
50%           NaN             6.000000       NaN         NaN   
75%           NaN             9.000000       NaN         NaN   
max           NaN            12.000000       NaN         NaN   

       unrepaired_damange           ad_created  nr_of_pictures   postal_code  \
count               40171                50000         50000.0  50000.000000   
unique                  2                   76             NaN           NaN   
top                  nein  2016-04-03 00:00:00             NaN           NaN   
freq                35232                 1946             NaN           NaN   
mean                  NaN                  NaN             0.0  50813.627300   
std                   NaN                  NaN             0.0  25779.747957   
min                   NaN                  NaN             0.0   1067.000000   
25%                   NaN                  NaN             0.0  30451.000000   
50%                   NaN                  NaN             0.0  49577.000000   
75%                   NaN                  NaN             0.0  71540.000000   
max                   NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [5]:
autos['offer_type'].value_counts()

offer_type
Angebot    49999
Gesuch         1
Name: count, dtype: int64

# changes to be made
The following list tabulates what should be modified in cleaning of the data. 

- oodmeter has a Km written in the value and the rpice column has a $ sign (that will be transfered to column name and removed from the individual value in the column)
- 
- seller has no information contained as all but one value are listed privet. Either the information is irrelevant or not granualar enough to differentiate between the privet values.
-
- offer_type columns as well contains no real information all but one value is differnet.
- The rpice column has a weird value that there are 1421 values with $0 price. This might not be an anomaly as some cars might be acutally just donated and up for grabs by anyone free of cost.
- There seems to be a abtest column with an equal separation in a and b, I am unaware what this abtesting was (will read the documentation)
- Anomaly in the year of registration, one value is mislabled for sure as the vehicle registration was made in 1001, either it was 2001 or something else. but we will leave it as it is, but when we perform analysis on this column, we will remove that entry.
- there are about 20% values that do not have a registration month associated with them (value is set to 0)
- Unrepaired damage is a binary in german, we can convert the vlues to yes and no for ja and nein, if we want, but it is understandable for me.
- nr_of_pictures column is all 0 so useless, toss it out.

Every other column seems to be good. 

In [6]:
# remove non numeric values in columns of price and odometer

autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype('int64')

autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype('int64')


new_col_dict = {
    'price' : 'price_in_dollar', 
    'odometer' : 'odometer_km'
}
autos = autos.rename(columns = new_col_dict)


In [7]:
# remove the colums that contain invariant information

autos.drop(['seller', 'offer_type', 'nr_of_pictures'], axis =1).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   date_crawled         50000 non-null  object
 1   name                 50000 non-null  object
 2   price_in_dollar      50000 non-null  int64 
 3   abtest               50000 non-null  object
 4   vehicle_type         44905 non-null  object
 5   registration_year    50000 non-null  int64 
 6   gearbox              47320 non-null  object
 7   power_ps             50000 non-null  int64 
 8   model                47242 non-null  object
 9   odometer_km          50000 non-null  int64 
 10  registratioon_month  50000 non-null  int64 
 11  fuel_type            45518 non-null  object
 12  brand                50000 non-null  object
 13  unrepaired_damange   40171 non-null  object
 14  ad_created           50000 non-null  object
 15  postal_code          50000 non-null  int64 
 16  last

In [17]:
print(
    autos['price_in_dollar'].unique().shape,
    autos['price_in_dollar'].describe(),
    autos['price_in_dollar'].value_counts().sort_index().head(20)
)

(2347,) count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price_in_dollar, dtype: float64 price_in_dollar
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: count, dtype: int64


In [9]:
print(
    autos['odometer_km'].unique().shape,
    autos['odometer_km'].describe(),
    autos['odometer_km'].value_counts().sort_index()
)

(13,) count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 odometer_km
5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: count, dtype: int64


# Issues with the price column, outliers
There are certain values that are too many. I decided to remove all the values that are greater than 900,000 as those are more than an order of magnitude higher than the previous highest value. 

There remain alot of values that are 0 for the price, I believe that is possible if the car is truely worthless. So I let that be in the dataset. 


In [18]:
autos = autos.loc[autos['price_in_dollar'] < 900000]

In [20]:
autos['date_crawled'].describe()

count                   49986
unique                  48200
top       2016-04-02 15:49:30
freq                        3
Name: date_crawled, dtype: object

In [46]:
print(
    autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(),
    autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(),
    autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(),
    autos['registration_year'].describe(include='all'),
    autos['registration_year'].value_counts().sort_index()
)

date_crawled
2016-03-05    0.025262
2016-03-06    0.014079
2016-03-07    0.036155
2016-03-08    0.033510
2016-03-09    0.033364
2016-03-10    0.032198
2016-03-11    0.032364
2016-03-12    0.036696
2016-03-13    0.015745
2016-03-14    0.036405
2016-03-15    0.034030
2016-03-16    0.029386
2016-03-17    0.031698
2016-03-18    0.012975
2016-03-19    0.034801
2016-03-20    0.037967
2016-03-21    0.037467
2016-03-22    0.032781
2016-03-23    0.032343
2016-03-24    0.029261
2016-03-25    0.031614
2016-03-26    0.032302
2016-03-27    0.030761
2016-03-28    0.034613
2016-03-29    0.034155
2016-03-30    0.033760
2016-03-31    0.031864
2016-04-01    0.033926
2016-04-02    0.035488
2016-04-03    0.038883
2016-04-04    0.036613
2016-04-05    0.012996
2016-04-06    0.003103
2016-04-07    0.001437
Name: proportion, dtype: float64 ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.039112


In [45]:
autos = autos[autos['registration_year'].between(1900, 2016)]

# remove outliers from the registration data
Remove vehicles that are registed after they were listed on the ebay for sale, and remove ancient values, before cars were athing i.e. before 20th century
We removed a lot of values by this filtering, final entries reomved now stand at about 2000. 

In [51]:
print(
    autos['brand'].unique().shape,
    autos['brand'].describe(),
    autos['brand'].value_counts().sort_values()
)

(40,) count          48016
unique            40
top       volkswagen
freq           10185
Name: brand, dtype: object brand
lada                 29
lancia               52
rover                65
daewoo               72
trabant              75
jaguar               76
saab                 77
land_rover           98
subaru              105
jeep                108
daihatsu            123
dacia               123
chrysler            176
chevrolet           274
suzuki              284
porsche             293
alfa_romeo          318
kia                 341
honda               377
mitsubishi          391
mini                415
volvo               444
hyundai             473
sonstige_autos      523
toyota              599
citroen             668
smart               668
nissan              725
mazda               727
skoda               770
seat                873
fiat               1242
peugeot            1418
renault            2274
ford               3350
audi               4149
mercedes_benz

In [55]:
brand_index = autos['brand'].value_counts().index
print(brand_index.shape)

(40,)


In [69]:
mean_price_brand = {}

for value in brand_index[0:8]:
    subset_data = autos.loc[autos['brand'] == value] 
    mean_price_brand[value] = subset_data['price_in_dollar'].mean()

print(mean_price_brand)


{'volkswagen': 5231.081983308787, 'bmw': 8102.536248343744, 'opel': 2876.716403542549, 'mercedes_benz': 8485.239571958942, 'audi': 9093.65003615329, 'ford': 3652.095223880597, 'renault': 2395.4164467897976, 'peugeot': 3039.4682651622}


In [71]:
mean_price = {}
mean_milage = {}
mean_registration = {}

for value in brand_index[0:8]:
    subset_data = autos.loc[autos['brand'] == value] 
    mean_price[value] = subset_data['price_in_dollar'].mean()
    mean_milage[value] = subset_data['odometer_km'].mean()
    mean_registration[value] = subset_data['registration_year'].mean()

print(mean_price, mean_milage, mean_registration)

{'volkswagen': 5231.081983308787, 'bmw': 8102.536248343744, 'opel': 2876.716403542549, 'mercedes_benz': 8485.239571958942, 'audi': 9093.65003615329, 'ford': 3652.095223880597, 'renault': 2395.4164467897976, 'peugeot': 3039.4682651622} {'volkswagen': 128724.10407461954, 'bmw': 132431.38368351315, 'opel': 129223.14208702349, 'mercedes_benz': 130856.0821139987, 'audi': 129287.78018799711, 'ford': 124068.65671641791, 'renault': 128183.81706244503, 'peugeot': 127136.81241184767} {'volkswagen': 2002.3598429062347, 'bmw': 2002.9333711906113, 'opel': 2002.140931844436, 'mercedes_benz': 2001.980345053505, 'audi': 2004.070378404435, 'ford': 2002.6597014925374, 'renault': 2002.8232189973614, 'peugeot': 2003.7242595204514}


In [79]:
mean_price_df = pd.Series(mean_price).sort_values(ascending= False)
mean_milage_df = pd.Series(mean_milage)
mean_registration_df = pd.Series(mean_registration)
df = pd.DataFrame(mean_price_df, columns = ['mean_price'])
df['mean_milage'] = mean_milage_df
df['mean_registration'] = mean_registration_df
print(df)

                mean_price    mean_milage  mean_registration
audi           9093.650036  129287.780188        2004.070378
mercedes_benz  8485.239572  130856.082114        2001.980345
bmw            8102.536248  132431.383684        2002.933371
volkswagen     5231.081983  128724.104075        2002.359843
ford           3652.095224  124068.656716        2002.659701
peugeot        3039.468265  127136.812412        2003.724260
opel           2876.716404  129223.142087        2002.140932
renault        2395.416447  128183.817062        2002.823219


# anaysis

While the comapnies audi, mercedes, and bmw has the highest resale value, we arent sure what the buying price of these was, if we knew that, we could have calculated the % deprecation.

What is interesting is that the almost all the cars milage is pretty compareable.

Hoever it seems like the audi cars were on the market the fastest, while the mercedes were retained by the owners for the longest time. That does not however correlate with the milage, which is pretty comparable for all the car brands. 


# Next steps

Data cleaning next steps:
- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
- See if there are particular keywords in the name column that you can extract as new columns

Analysis next steps:
- Find the most common brand/model combinations
- Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
- How much cheaper are cars with damage than their non-damaged counterparts?